CSS :state() pseudo class (#32000)
mdn/content: files/en-us/web/api/customstateset/index.md

### Interaction with CSS

> Developers can select a custom element with a specific state using the [`:state()`](/en-US/docs/Web/CSS/:state) _custom state pseudo-class_.

Author: I tend to use "you" but never entirely sure. FYI I have been using the "with a specific state" because I've been thinking of the state as some data owned by the element. That's not a useful mental model, so thanks for fixing this.

```suggestion
You can select a custom element that is in a specific state using the [`:state()`](/en-US/docs/Web/CSS/:state) _custom state pseudo-class_.
```

The format of this pseudo-class is `:state(mystatename)`, where `mystatename` is the state as defined in the element.
The custom state pseudo-class matches the custom element only if the state is `true` (i.e. if `mystatename` is present in the `CustomStateSet`).

> For example, the following CSS might be used to match a `labeled-checkbox` custom element where the element's `CustomStateSet` contains `checked`, and apply a solid border:

Author: Slight modification

```suggestion
For example, the following CSS matches a `labeled-checkbox` custom element when the element's `CustomStateSet` contains the `checked` state, and applies a `solid` border to the checkbox:
```

> For example, the following CSS might be used to match a `labeled-checkbox` custom element where the element's `CustomStateSet` contains `checked`, and apply a solid border:

Reviewer: This is good, thanks

```css
labeled-checkbox:state(checked) {
  border: solid;
}
```

CSS can also match a custom state [within custom element's shadow DOM](/en-US/docs/Web/CSS/:state#matching_a_custom_state_in_a_custom_elements_shadow_dom) by specifying `:state()` within the [`:host()`](/en-US/docs/Web/CSS/:host_function) pseudo-class function.

`:state()` can be used after the [`::part()`](/en-US/docs/Web/CSS/::part) pseudo element to match the [shadow parts](/en-US/docs/Web/CSS/CSS_shadow_parts) of a custom element that have a particular state.

> **Warning:** Chrome supports a deprecated syntax that selects custom states using a CSS `<dashed-ident>` rather than the `:state()` function.
> For information about how to support both approaches see the [Compatibility with `<dashed-ident>` syntax](compability_with_dashed-ident_syntax) section below.

## Examples

> ### Labeled checkbox

Author: It is good to fix now. Done in https://github.com/mdn/content/pull/32000/commits/12d3853e225af472515309dc9623a7bc52c4c860

{{EmbedLiveSample("Labeled Checkbox", "100%", 50)}}

> ### Question box: exposing states in shadow parts

Author: Good point

```suggestion
### Matching a custom state in the shadow parts of a custom element
```

         white-space: pre;
         font-family: monospace;
       }
       :host(:state(checked))::before { content: '[x]'; background: grey; }
       </style>
       <slot>Label</slot>`;
  }

  get checked() {
    return this._internals.states.has("checked");
  }

  set checked(flag) {
    if (flag) {
      this._internals.states.add("checked");
    } else {
      this._internals.states.delete("checked");
    }
  }

  _onClick(event) {
    // Toggle the 'checked' property when the element is clicked
    this.checked = !this.checked;
  }
}

customElements.define("labeled-checkbox", LabeledCheckbox);
```

First we define our custom element class `QuestionBox`, which extends `HTMLElement`.
As always, the constructor first calls the `super()` method.
It then calls [`attachShadow()`](/en-US/docs/Web/API/Element/attachShadow) to attach a shadow DOM tree to the custom element.
The shadow DOM is attached with `mode: 'closed'` to ensure that nodes within the shadow DOM are not available outside of the element, unless explicitly exposed as a [shadow part](/en-US/docs/Web/CSS/CSS_shadow_parts).

```js
class QuestionBox extends HTMLElement {
  constructor() {
    super();

>     const shadowRoot = this.attachShadow({ mode: "closed" });

Reviewer 2: The difference between an open vs closed shadowroot is quite small, so the meaningful benefits of encapsulation are already provided with an open shadowroot. Closed shadowroots are ever so slightly stricter, but can cause surprising ramifications for the developer. The biggest is that `el.shadowRoot` will be `null` even though the element does have a ShadowRoot. Some smaller ramifications are that `.composedPath()` on composed events emitted from within the shadowroot will elide shadow elements, and `.assignedSlot` on an element will always be `null` even if that element is assigned to a closed shadowroot slot. These seem like very small differences but can cause developer confusion unless the developer is keenly aware of these changes. Given that, my opinion is that unless the example is specifically requiring a closed shadowroot (for example for the purposes of demonstrating the above) it should default to `"open"` so that examples that get copy/pasted don't end up hitting these gotchas.

>     const shadowRoot = this.attachShadow({ mode: "closed" });

Author: @keithamus Thanks for that explanation. For this example I am happy to change it.

As far as I can tell little of the above is explained in https://developer.mozilla.org/en-US/docs/Web/API/Element/attachShadow - which is the point you would make the decision. It is spread around in https://developer.mozilla.org/en-US/docs/Web/API/Element/shadowRoot (fact that this is `null` if mode was attached as closed), https://developer.mozilla.org/en-US/docs/Web/API/Event/composedPath() - etc.

Do you think you might be able to add some reasoning in https://developer.mozilla.org/en-US/docs/Web/API/Element/attachShadow as you have done here to explain this. I would expect that to cover why you would choose to use `closed` - I mean the spec authors must have seen it as useful at some point?

It then defines a `<labeled-checkbox>` custom element with default text`"Yes"`.
This checkbox is exposed as a shadow part of the question box with the name `checkbox` using the [`part`](/en-US/docs/Web/HTML/Global_attributes#part) attribute.

Note that the code and styling for the `<labeled-checkbox>` is exactly the same as in [previous example](#labeled_checkbox), and is therefore not shown.

We then call the {{domxref("CustomElementRegistry/define", "define()")}} method on the object returned by {{domxref("Window.customElements")}} in order to register the custom element with name `<question-box>`:

```js
customElements.define("question-box", QuestionBox);
```

#### HTML

After registering the custom element we can use the element in HTML as shown below.

```html
<!-- Question box with default prompt "Question" -->
<question-box></question-box>
<!-- Question box with custom prompt "Continue?" -->
<question-box>Continue?</question-box>
```

#### CSS

The first block of CSS uses the [`::part()`](/en-US/docs/Web/CSS/::part) selector to match on the exposed shadow part named `checkbox`, setting it to be red by default.

```css
question-box::part(checkbox) {
  color: red;
}
```

> We then use `:state()` after `::part()` to match on `checkbox` parts with the `checked` state:

Author: Don't know why, but I find that hard to parse. Compromise.

```suggestion
The second block follows `::part()` with `:state()` to match the `checkbox` parts that are in the `checked` state:
```

    this._internals = this.attachInternals();
    this.addEventListener("click", this._onClick.bind(this));

    const shadowRoot = this.attachShadow({ mode: "open" });
    shadowRoot.innerHTML = `<style>
        :host {
          display: block;
        }
       :host::before {
         content: '[ ]';
         white-space: pre;
         font-family: monospace;
       }
       :host(:state(checked))::before { content: '[x]'; }
       </style>
       <slot>Label</slot>`;
  }

  get checked() {
    return this._internals.states.has("checked");
  }

  set checked(flag) {
    if (flag) {
      this._internals.states.add("checked");
    } else {
      this._internals.states.delete("checked");
    }
  }

  _onClick(event) {
    // Toggle the 'checked' property when the element is clicked
    this.checked = !this.checked;
  }
}

customElements.define("labeled-checkbox", LabeledCheckbox);
```

First, we define the custom element class `QuestionBox`, which extends `HTMLElement`.
As always, the constructor first calls the `super()` method.
Next, we attach a shadow DOM tree to the custom element by calling [`attachShadow()`](/en-US/docs/Web/API/Element/attachShadow).

> The shadow DOM is attached with `mode: 'closed'` to ensure that nodes within the shadow DOM are not available from outside the element, unless explicitly exposed as a [shadow part](/en-US/docs/Web/CSS/CSS_shadow_parts).

Author: Good catch. I'm just going to remove this line as it is no longer needed if we're not going to hide the shadow dom.

```suggestion
```
